addons/l10n_in_edi/models/account_edi_format.py

```diff
@@ -0,0 +1,442 @@
+                    (tax_details.get("base_amount_currency") + tax_details.get("tax_amount_currency")) * sign, 2)
+            })
+        if saler_buyer.get("seller_details") != saler_buyer.get("dispatch_details"):
+            json_payload.update({
+                "DispDtls": self._get_l10n_in_edi_partner_details(move, saler_buyer.get("dispatch_details"),
+                    vat=False, phone_and_email=False)
+            })
+        if saler_buyer.get("buyer_details") != saler_buyer.get("ship_to_details"):
+            json_payload.update({
+                "ShipDtls": self._get_l10n_in_edi_partner_details(move, saler_buyer.get("ship_to_details"))
+            })
+        if move.l10n_in_gst_treatment == "overseas":
+            json_payload.update({
+                "ExpDtls": {
+                    "ShipBNo": move.l10n_in_shipping_bill_number or "",
+                    "ShipBDt": move.l10n_in_shipping_bill_date
+                       and move.l10n_in_shipping_bill_date.strftime("%d/%m/%Y") or "",
+                    "Port": move.l10n_in_shipping_port_code_id.code or "",
+                    "RefClm": tax_details_by_code.get("igst") and "Y" or "N",
+                    "ForCur": move.currency_id.name,
+                    "CntCode": saler_buyer.get("buyer_details").country_id.code or "",
+                }
+            })
+        return json_payload
+
+    @api.model
+    def _l10n_in_grouping_key_generator(self, tax_values):
+        base_line = tax_values["base_line_id"]
+        tax_line = tax_values["tax_line_id"]
+        line_code = "other"
+        if tax_line.tax_tag_ids in self.env.ref("l10n_in.tax_report_line_cess").tag_ids:
+            if tax_line.tax_line_id.amount_type != "percent":
+                line_code = "cess_non_advol"
+            else:
+                line_code = "cess"
+        elif tax_line.tax_tag_ids in self.env.ref("l10n_in.tax_report_line_state_cess").tag_ids:
+            if tax_line.tax_line_id.amount_type != "percent":
+                line_code = "state_cess_non_advol"
+            else:
+                line_code = "state_cess"
+        else:
+            for gst in ["cgst", "sgst", "igst"]:
+                if tax_line.tax_tag_ids in self.env.ref("l10n_in.tax_report_line_%s"%(gst)).tag_ids:
+                    line_code = gst
+        return {
+            "tax": tax_values["tax_id"],
+            "base_product_id": base_line.product_id,
+            "tax_product_id": tax_line.product_id,
+            "base_product_uom_id": base_line.product_uom_id,
+            "tax_product_uom_id": tax_line.product_uom_id,
+            "line_code": line_code,
+        }
+
+    @api.model
+    def _l10n_in_filter_to_apply(self, tax_values):
+        if tax_values["base_line_id"].is_rounding_line:
+            return False
+        return True
+
+    @api.model
+    def _l10n_in_prepare_edi_tax_details(self, move, in_foreign=False):
+        invoice_tax_values = move._prepare_edi_tax_details(
```

Comment:
Plz avoid flooding with a lot of sub-methods. This is not making the code cleaner. You can do it directly inside the "main" json method. It's only my opinion but I think this is more readable to have one big method doing all the trick but splitted with some comments like:
https://github.com/odoo/enterprise/blob/master/l10n_mx_edi/models/account_edi_format.py#L233

Also, suppose the api chance one day. This will be a real pain in the ass to fight with the whole stable policy by adding extra parameters everywhere if the code is spread into a lot of tiny sub-methods.